### ISSUE TITLE
[Question] Clarification on `member_ordering` configuration

### ISSUE THREAD

**Issue by GroovinChip** [dart-code-checker/dart-code-metrics#242]
A `member_ordering` configuration with constructors declared before public fields seems to be incorrectly warning that public fields should be before constructors. Is this correct behavior or a bug?
 
Configuration in `analysis_options.yaml`:
```yaml
...
dart_code_metrics:
  ...
  member-ordering:
    alphabetize: true
    order:
      - constructors
      - public_fields
      - private_fields
```
Screenshot of warning:
<img width="444" alt="Screen Shot 2021-04-02 at 12 36 36 PM" src="https://user-images.githubusercontent.com/4250470/113435065-19203380-93b0-11eb-8b52-0c8ac5bf93ef.png">

Code in question:
```dart
class CallManagerApp extends StatefulWidget {
  const CallManagerApp({
    Key key,
    @required this.prefsService,
    @required this.contactsUtility,
    @required this.phoneUtility,
    @required this.notificationService,
  }) : super(key: key);

  final PrefsService prefsService; // warning on this line
  final ContactsUtility contactsUtility; // warning on this line
  final PhoneUtility phoneUtility; // warning on this line
  final NotificationService notificationService; // warning on this line

  @override
  _CallManagerAppState createState() => _CallManagerAppState();
}
```

**Comment by incendial** [comment]
@GroovinChip looks like a bug to me, but is it a full rule config? And can you share a full content of the file? Also is it a correct screenshot, because it should not warn about methods for that case?

**Comment by GroovinChip** [comment]
No, it's not a full rule config. The way I understood the [docs](https://github.com/dart-code-checker/dart-code-metrics/blob/master/doc/rules/member_ordering.md) was that I could choose specific ordering rules and anything not specified would fall under the default ordering rule.

Here's the full file:
```yaml
linter:
  rules:
    - prefer_single_quotes
    - unnecessary_new
    - prefer_contains
    - package_names
    - non_constant_identifier_names
    - library_prefixes
    - library_names
    - implementation_imports
    - file_names
    - camel_case_extensions
    - camel_case_types
    - avoid_empty_else
    - avoid_classes_with_only_static_members
    - always_declare_return_types
    - always_require_non_null_named_parameters

analyzer:
  exclude:
    - ios/**
  plugins:
    - dart_code_metrics

dart_code_metrics:
  anti-patterns:
    - long-method
    - long-parameter-list
  metrics:
    cyclomatic-complexity: 20
    lines-of-executable-code: 50
    number-of-arguments: 4
    maximum-nesting: 5
  metrics-exclude:
    - test/**
  rules:
    - newline-before-return
    - no-boolean-literal-compare
    - no-empty-block
    - prefer-trailing-comma
    - no-equal-then-else
    - potential-null-dereference
  member-ordering:
    alphabetize: true
    order:
      - constructors
      - public_fields
      - private_fields
```

**Comment by incendial** [comment]
Sorry, I meant the dart file content.

> The way I understood the docs was that I could choose specific ordering rules and anything not specified would fall under the default ordering rule.

Thats correct. But the error still looks strange, because methods are not mentioned in the config at all.

**Comment by incendial** [comment]
Right now we don't support auto config update applying, analysis server (or simply IDE) should be reloaded after config change. Can that be the issue for your case?

**Comment by GroovinChip** [comment]
> Sorry, I meant the dart file content.

Oh, sorry about that, here it is:
```dart
import 'package:call_manager/firebase/firebase.dart';
import 'package:call_manager/screens/home_screen.dart';
import 'package:call_manager/screens/login_screen.dart';
import 'package:call_manager/services/contacts_utility.dart';
import 'package:call_manager/services/notifications_service.dart';
import 'package:call_manager/services/phone_utility.dart';
import 'package:call_manager/services/prefs_service.dart';
import 'package:call_manager/theme/app_themes.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

class CallManagerApp extends StatefulWidget {
  const CallManagerApp({
    Key key,
    @required this.prefsService,
    @required this.contactsUtility,
    @required this.phoneUtility,
    @required this.notificationService,
  }) : super(key: key);

  final PrefsService prefsService;
  final ContactsUtility contactsUtility;
  final PhoneUtility phoneUtility;
  final NotificationService notificationService;

  @override
  _CallManagerAppState createState() => _CallManagerAppState();
}

class _CallManagerAppState extends State<CallManagerApp> with FirebaseMixin {
  final _navigatorKey = GlobalKey<NavigatorState>();

  @override
  void initState() {
    super.initState();
    _onAuthStateChange();
  }

  void _onAuthStateChange() {
    auth.authStateChanges().listen((User user) {
      if (user == null) {
        _navigatorKey.currentState.pushAndRemoveUntil(
          LoginScreen.route(),
          (route) => false,
        );
      } else {
        firestore.initStorageForUser(currentUser.uid);
        _navigatorKey.currentState.pushAndRemoveUntil(
          HomeScreen.route(),
          (route) => false,
        );
      }
    });
  }

  Route<dynamic> _onGenerateRoute(RouteSettings settings) {
    switch (settings.name) {
      case HomeScreen.routeName:
        return HomeScreen.route(settings: settings);
      case LoginScreen.routeName:
        return LoginScreen.route(settings: settings);
      default:
        return null;
    }
  }

  @override
  Widget build(BuildContext context) {
    return MultiProvider(
      providers: [
        Provider<PrefsService>.value(value: widget.prefsService),
        Provider<ContactsUtility>.value(value: widget.contactsUtility),
        Provider<PhoneUtility>.value(value: widget.phoneUtility),
        Provider<NotificationService>.value(value: widget.notificationService),
      ],
      child: StreamBuilder<ThemeMode>(
        stream: widget.prefsService.themeModeSubject,
        initialData: widget.prefsService.themeModeSubject.valueWrapper.value,
        builder: (context, snapshot) {
          return MaterialApp(
            navigatorKey: _navigatorKey,
            title: 'Call Manager',
            theme: AppThemes.lightTheme(),
            darkTheme: AppThemes.darkTheme(),
            themeMode: snapshot.data,
            initialRoute: currentUser != null
                ? HomeScreen.routeName
                : LoginScreen.routeName,
            onGenerateInitialRoutes: (String initialRoute) => [
              _onGenerateRoute(RouteSettings(name: initialRoute)),
            ],
            onGenerateRoute: _onGenerateRoute,
            debugShowCheckedModeBanner: false,
          );
        },
      ),
    );
  }
}
```
> Right now we don't support auto config update applying, analysis server (or simply IDE) should be reloaded after config change. Can that be the issue for your case?
I guess it's possible, I didn't realize the analysis server needed to be restarted after changing my `dart_code_metrics` configuration. I restarted the server while typing this and it's taking ages - it's been pretty slow lately in general.

**Comment by incendial** [comment]
> I didn't realize the analysis server needed to be restarted after changing my dart_code_metrics configuration

We have a task to change the plugin to automatically update config after config file is changed and hopefully will take it soon.

> Edit 2: Actually it's back now. But the warnings have all disappeared

All wrong ones or just all?

**Comment by GroovinChip** [comment]
> All wrong ones or just all?

I mean the analyzer is working but now there are no warnings related to `dart_code_metrics` at all.

**Comment by incendial** [comment]
Hm, that's strange... I've cloned you repo and tested it with null safety metrics version, all works fine and event member ordering issues are shown.

If you try to run cli `pub run dart_code_metrics:metrics lib`, will it display member_ordering issues?

**Comment by GroovinChip** [comment]
Doing `flutter pub run dart_code_metrics:metrics lib` does nothing, actually. No output on the console. Maybe I need to update the global package.

**Comment by incendial** [comment]
Okey, I see it now. Your config should look like:
```yaml
...
  rules:
    - newline-before-return
    - no-boolean-literal-compare
    - no-empty-block
    - prefer-trailing-comma
    - no-equal-then-else
    - potential-null-dereference
    - member-ordering:
        alphabetize: true
        order:
          - constructors
          - public_fields
          - private_fields
```

**Comment by GroovinChip** [comment]
The tool is telling me that's not correct syntax 🤔

**Comment by GroovinChip** [comment]
Okay, things are working as expected now. Thank you for clearing this up!

**Comment by incendial** [comment]
No problem, thanks for the great feedback. I've added a task to make config examples more clear.
